include/beman/any_view/detail/unreachable_iterator.hpp

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef BEMAN_ANY_VIEW_DETAIL_UNREACHABLE_ITERATOR_HPP
+#define BEMAN_ANY_VIEW_DETAIL_UNREACHABLE_ITERATOR_HPP
+
+#include <beman/any_view/detail/utility.hpp>
+
+#include <compare>
+#include <iterator>
+
+namespace beman::any_view::detail {
+
+template <class ElementT, class RefT, class RValueRefT, class DiffT>
+struct iterator_concept {
+    using type = std::random_access_iterator_tag;
+};
+
+template <class ElementT>
+struct iterator_concept<ElementT, ElementT&, ElementT&&, std::ptrdiff_t> {
+    using type = std::contiguous_iterator_tag;
+};
+
+template <class ElementT, class RefT, class RValueRefT, class DiffT>
+class unreachable_iterator {
+  public:
+    using iterator_concept = iterator_concept<ElementT, RefT, RValueRefT, DiffT>::type;
+    using element_type     = ElementT;
+    using difference_type  = DiffT;
+
+    constexpr auto operator*() const noexcept -> RefT { unreachable(); }
+
+    constexpr friend auto iter_move(unreachable_iterator) noexcept -> RValueRefT { unreachable(); }
+
+    constexpr auto operator++() noexcept -> unreachable_iterator& { return *this; }
+
+    constexpr auto operator++(int) noexcept -> unreachable_iterator { return *this; }
+
+    constexpr auto operator<=>(const unreachable_iterator&) const noexcept -> std::strong_ordering = default;
+
+    constexpr auto operator--() noexcept -> unreachable_iterator& { return *this; }
+
+    constexpr auto operator--(int) noexcept -> unreachable_iterator { return *this; }
+
+    constexpr auto operator-(unreachable_iterator) const noexcept -> difference_type { return 0; }
+
+    constexpr auto operator+=(difference_type) noexcept -> unreachable_iterator& { return *this; }
+
+    constexpr auto operator+(difference_type) const noexcept -> unreachable_iterator { return *this; }
+
+    constexpr friend auto operator+(difference_type, unreachable_iterator) noexcept -> unreachable_iterator {
```

Comment:
I'm curious on the decision of which of these should be friends?

Reply:
`iter_move(it)` and `n + it` are declared as `friend`s in order to be called as free functions rather than member functions. Or was there something else you wanted clarified?

Reply:
> `iter_move(it)` and `n + it` are declared as `friend`s in order to be called as free functions rather than member functions. Or was there something else you wanted clarified?

Well operators<=> and such are mostly declared as friends in ranges as well -- so it struck me as different that these weren't.  Random example: https://eel.is/c++draft/range.transform.iterator

I guess though given that this is in detail this will be an expostion-only thing?

Reply:
This isn't even exposition, it's completely internal. Defining the operators as non-friends is more concise, but I gave no other consideration than that to the decision.

Disregarding the fact this is purely an implementation detail, is there anything significant that distinguishes the characteristics of non-friends from their friend counterparts when declaring these operators? I'm asking so I can be aware of any other possible considerations in the future.

Reply:
Oh, it's basically down to ADL - if the operators are friends they can be found by ADL, otherwise they're just members that can't be found that way.  Since it's an implementation detail it doesn't matter here of course.

Reply:
> if the operators are friends they can be found by ADL

Given that `x <=> y` can de-sugar to either `x.operator<=>(y)` or `operator<=>(x, y)`, does the ability to be found by ADL ever affect usage of the syntactic sugar, or only the ability to call it with the latter de-sugared syntax?